src/parser/expr.rs

```diff
@@ -27,90 +145,101 @@ impl<'s, 'a> Parser<'s, 'a> {
-            let span = self.spans.merge(&left, &right);
-            left = ExprKind::BinOp(BinOp::LogicalAnd, self.alloc(left), self.alloc(right))
-                .with_span(span);
-        }
-        Ok(left)
-    }
+    fn binary_expr(
+        &mut self,
+        mut lhs: Expr<'s, 'a>,
+        last_operator: Option<&Spanned<BinaryOp>>,
+    ) -> ParseResult<Expr<'s, 'a>> {
+        let precedence_bound = match last_operator {
+            Some(last_operator) => last_operator.precedence(),
+            None => Precedence::Lowest,
+        };
 
-    fn comparison(&mut self) -> ParseResult<Expr<'s, 'a>> {
-        let left = self.sum()?;
-        if let Some((op, _span)) = self.accept_any([
-            (tok![==], BinOp::Eq),
-            (tok![!=], BinOp::Neq),
-            (tok![>=], BinOp::Gte),
-            (tok![<=], BinOp::Lte),
-            (tok![>], BinOp::Gt),
-            (tok![<], BinOp::Lt),
-        ])? {
-            let right = self.sum()?;
-            let span = self.spans.merge(&left, &right);
-            return Ok(ExprKind::BinOp(op, self.alloc(left), self.alloc(right)).with_span(span));
-        } else {
-            Ok(left)
+        while let Some(operator) = self.peek_binary_operator()? {
+            // now we look at the precedence and associativity of our operator,
+            // and determine whether we should continue parsing more expression
+            // or return back up
+            match operator.precedence().compatibility(&precedence_bound) {
+                Compatibility::Continue => {
+                    // we peeked already that an operator is coming, we just need to know its span and progress the parser
+                    let (_, span) = self.next()?;
+                    let span = self.spans.store(span);
+                    let operator = operator.with_span(span);
+
+                    // if we continue,
+                    // parse the left hand side of the next expression
+                    let new_lhs = self.unary()?;
+                    // and now parse a next expression, with this newly parsed
+                    // left hand side. Maybe, it parses an operator too many,
+                    // in that case it returns it and we cache it in `maybe_operator`
+                    // and use it next if we determine we should not return it ourselves
+                    let rhs = self.binary_expr(new_lhs, Some(&operator))?;
+
+                    // make an expression, and loop
+                    let span = self.spans.merge(&lhs, &rhs);
+                    lhs = ExprKind::BinaryOp(operator, self.alloc(lhs), self.alloc(rhs))
+                        .with_span(span);
+                }
+                Compatibility::Stop => {
+                    return Ok(lhs);
+                }
+                Compatibility::Incompatible => {
+                    let last_operator = last_operator
+                        .ice("we can never have incompatibility if there was no last operator");
```

Comment:
We can avoid this unwrap with some light restructuring.  `Stop` and `Incompatible` only occur if `last_operator` was `Some`; so put this entire `match` inside an `if let` that unwraps `last_operator`.  In the `Continue` case, exit the `if let` and do all the setup work after it.  I think it'll actually improve the code structure, if the other suggestions are implemented as well.

Reply:
we can! and we did